### ISSUE TITLE
Training with Keras.metrics is giving the correct number of true negatives and false negatives, but 0 for true positives and false positives

### ISSUE THREAD

**Issue by drskaf** [keras-team/keras#18044]
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):
- Python version:
- Bazel version (if compiling from source):
- GPU model and memory:
- Exact command to reproduce:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"

**Describe the problem**.

Describe the problem clearly here. Be sure to convey here why it's a bug in Keras or why the requested feature is needed.

**Describe the current behavior**.

**Describe the expected behavior**.

I have imbalanced data with images and their binary classes (ratio of pos/neg is 6%). I loaded my images as list of images, binary labels from a data frame as a list of values, converted both to numpy array, then used few metrics from Keras.metrics trying to monitor precision. I followed tutorial from Tensorflow on 'credit fraud detection', I used exactly the activation function 'sigmoid', loss function 'BinaryCrossentropy', but I have been getting 0 values for true positive and false positives, so precision can't be calculated. I get it the other way round if I choose 'softmax' activation, I will appreciate some help. Thanks. 

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

This is a link to my GitHub repository, the code I'm working on is in file 'test_keras.py'
To reproduce the issue I will need to provide the images, but they are large dataset medical images with ethical approval restriction for research, so I wonder if advice can be given without sending the images across:

https://github.com/drskaf/Stress-Perf-Classifier.git

**Source code / logs**.

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

This is a snippet of the training log:

Epoch 1/100
66/66 [==============================] - 73s 1s/step - loss: 0.8304 - tp: 0.0000e+00 - fp: 0.0000e+00 - tn: 1987.0000 - fn: 113.0000 - accuracy: 0.9462 - precision: 0.0000e+00 - recall: 0.0000e+00 - auc: 0.5000 - prc: 0.0538 - val_loss: 0.8227 - val_tp: 0.0000e+00 - val_fp: 0.0000e+00 - val_tn: 852.0000 - val_fn: 48.0000 - val_accuracy: 0.9467 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00 - val_auc: 0.5000 - val_prc: 0.0533

**Comment by Frightera** [comment]
Hi @drskaf,

It looks like your model is just outputting zeros everytime. If you look at the accuracy:

`1987 / 2100 = 0.9462`

Basically everything is classified as class 0, that's why you see` tp: 0.0000e+00 - fp: 0.0000e+00 - tn: 1987.0000 - fn: 113.0000`.

**Comment by drskaf** [comment]
Thanks,
But why is this behaviour? 
I have label arrays of [0]s and [1]s, I choose BinaryCrossentropy as per tutorial, and 'sigmoid' function, the output of the model predicts series of arrays in the shape of [float], but Keras.metrics outputting only zeros as you mentioned. Does that mean I have to make my labels as one array per batch size instead of series of arrays to make the metrics work on both zeros and ones?

**Comment by Frightera** [comment]
@drskaf 

This is expected behaviour as model do not predict class 1, the predicted classes by model are just class 0s. Might be related to model or data, you need to consider different strategies for training.

**Comment by drskaf** [comment]
After changing my image loading function to make images a column in the data frame containing the labels, then creating 2d NumPy array from the images using: X_train = np.array([np.array(x) for x in df_train['images']]), and taking the labels out of the data frame: y_train = np.array(df_train.pop(args["target"])) I managed to get all tp, fp, tn, fn in the first epoch, but not been able to get them on the validation set or from 2nd epoch onwards. I will be grateful for any comments as I don't understand the behaviour of BinaryCrossentropy with my model? thanks

Here is a snippet of first 2 epochs:
Epoch 1/100
33/33 [==============================] - 70s 2s/step - loss: 0.5887 - tp: 4.0000 - fp: 60.0000 - tn: 1927.0000 - fn: 109.0000 - accuracy: 0.9195 - precision: 0.0625 - recall: 0.0354 - auc: 0.4962 - prc: 0.0543 - val_loss: 0.2195 - val_tp: 0.0000e+00 - val_fp: 0.0000e+00 - val_tn: 852.0000 - val_fn: 48.0000 - val_accuracy: 0.9467 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00 - val_auc: 0.5000 - val_prc: 0.0533
Epoch 2/100
33/33 [==============================] - 69s 2s/step - loss: 0.2204 - tp: 0.0000e+00 - fp: 0.0000e+00 - tn: 1987.0000 - fn: 113.0000 - accuracy: 0.9462 - precision: 0.0000e+00 - recall: 0.0000e+00 - auc: 0.5000 - prc: 0.0538 - val_loss: 0.2182 - val_tp: 0.0000e+00 - val_fp: 0.0000e+00 - val_tn: 852.0000 - val_fn: 48.0000 - val_accuracy: 0.9467 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00 - val_auc: 0.5098 - val_prc: 0.0676

**Comment by drskaf** [comment]
After waiting for more than 20 epochs, I was getting all values and the model just needed some time to learn, so I think all is good now! 
Thanks

**Comment by Frightera** [comment]
Hi @drskaf,

Kindly close the issue if your problem is solved.
